drivers/watchdog/wdt_npcx.c

```diff
@@ -0,0 +1,302 @@
+ *  |
+ *  |    +-------------------+    +-----------------+
+ *  +--->| Watchdog Prescale |--->| 8-Bit Watchdog  |-----> Watchdog Event/Reset
+ *       |    (WDCP 1:32)    |    | Counter (WDCNT) |       after n clocks
+ *       +-------------------+    +-----------------+
+ *
+ */
+
+#include <assert.h>
+#include <drivers/gpio.h>
+#include <drivers/clock_control.h>
+#include <drivers/watchdog.h>
+#include <soc.h>
+
+#include "soc_miwu.h"
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wdt_npcx, CONFIG_WDT_LOG_LEVEL);
+
+/* Watchdog operating frequency is fixed to LFCLK (32.768) kHz */
+#define NPCX_WDT_CLK LFCLK
+
+/*
+ * Maximum watchdog window time. Since the watchdog counter is 8-bits, maximum
+ * time supported by npcx watchdog is 256 * (32 * 32) / 32768 = 8 sec.
+ */
+#define NPCX_WDT_MAX_WND_TIME 8000UL
+
+/*
+ * Minimum watchdog window time. Ensure we have waited at least 3 watchdog
+ * clocks since touching WD timer. 3 / (32768 / 1024) HZ = 93.75ms
+ */
+#define NPCX_WDT_MIN_WND_TIME 100UL
+
+/* Device config */
+struct wdt_npcx_config {
+	/* wdt controller base address */
+	uintptr_t base;
+	/* t0 timer wake-up input source configuration */
+	const struct npcx_wui t0out;
+};
+
+/* Driver data */
+struct wdt_npcx_data {
+	int64_t last_watchdog_touch;
+	wdt_callback_t cb;
+	uint32_t timeout;
+	bool timeout_installed;
+};
+
+struct miwu_dev_callback miwu_cb;
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct wdt_npcx_config *)(dev)->config)
+#define DRV_DATA(dev) ((struct wdt_npcx_data *)(dev)->data)
+#define HAL_INSTANCE(dev) (struct twd_reg *)(DRV_CONFIG(dev)->base)
+
+/* WDT local functions */
+static void wdt_t0out_isr(const struct device *dev, struct npcx_wui *wui)
+{
+	struct wdt_npcx_data *const data = DRV_DATA(dev);
+	ARG_UNUSED(wui);
+
+	/* Handle watchdog event here. */
+	if (data->cb) {
+		data->cb(dev, 0);
+	}
+
+	LOG_DBG("WDT issued! WUI(%d %d %d)", wui->table, wui->group, wui->bit);
+
+	/* Wait for watchdog occurred! */
+	while (1)
+		;
+}
+
+static void wdt_config_t0out_interrupt(const struct device *dev)
+{
+	const struct wdt_npcx_config *const config = DRV_CONFIG(dev);
+
+	/* Initialize a miwu device input and its callback function */
+	npcx_miwu_init_dev_callback(&miwu_cb, &config->t0out, wdt_t0out_isr,
+			dev);
+	npcx_miwu_manage_dev_callback(&miwu_cb, true);
+
+	/*
+	 * Configure the T0 wake-up event triggered from a rising edge
+	 * on T0OUT signal.
+	 */
+	npcx_miwu_interrupt_configure(&config->t0out,
+			NPCX_MIWU_MODE_EDGE, NPCX_MIWU_TRIG_HIGH);
+}
+
+/* WDT api functions */
+static int wdt_npcx_install_timeout(const struct device *dev,
+				   const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_npcx_data *const data = DRV_DATA(dev);
+	struct twd_reg *const inst = HAL_INSTANCE(dev);
+
+	/* If watchdog is already running */
+	if (IS_BIT_SET(inst->T0CSR, NPCX_T0CSR_WD_RUN)) {
+		return -EBUSY;
+	}
+
+	/* No window watchdog support */
+	if (cfg->window.min != 0U) {
+		return -EINVAL;
+	}
+
```

Comment:
I feel it's too strict... This driver can accept windowed timeout (ex. min=1, max=10) parameters because it uses only max timeout.
And forgot to add "timeout_installed = false"??

Reply:
> I feel it's too strict... This driver can accept windowed timeout (ex. min=1, max=10) parameters because it uses only max timeout.
[ML] NPCX series doesn't support window watchdog. The checking in line 132 makes sure the preload value of watchdog counter register is within valid range as nrf driver did [here](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/watchdog/wdt_nrfx.c#L95). I will add more details on the comment of this line.

 > And forgot to add "timeout_installed = false"??
[ML] Thanks for reminding. I will add it later.

Reply:
Done.